File: Feed/src/main/java/com/j/jface/feed/JFaceDataFeed.java
package com.j.jface.feed;

import android.Manifest;
import android.app.Activity;
import android.app.Fragment;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import com.j.jface.R;
import com.j.jface.feed.fragments.DebugToolsFragment;
import com.j.jface.feed.fragments.ImageSelectorFragment;
import com.j.jface.feed.fragments.LogsAndDataFragment;
import com.j.jface.feed.fragments.MessagesFragment;
import com.j.jface.lifecycle.FragmentWrapper;
import com.j.jface.lifecycle.WrappedActivity;
import com.j.jface.wear.Wear;

public class JFaceDataFeed extends WrappedActivity
{
  @NonNull private final String LAST_OPEN_FRAGMENT_INDEX = "last_open_fragment_index";
  @NonNull private final ActionBarDrawerToggle mDrawerToggle;
  @NonNull private final Wear mWear;

  // State
  private int mCurrentlyDisplayedFragmentIndex = 0;

  public JFaceDataFeed(@NonNull final Args args)
  {
    super(args);
    mA.setContentView(R.layout.data_feed_drawer);
    mWear = new Wear(mA);
    final DrawerLayout drawer = mA.findViewById(R.id.dataFeedDrawer);

    final ListView list = mA.findViewById(R.id.dataFeedDrawerContents);
    list.setAdapter(new ArrayAdapter<>(mA, R.layout.data_feed_drawer_item, new String[] { "Messages", "Settings", "Logs & data", "Debug tools" }));
    final AdapterView.OnItemClickListener listener = (parent, view, position, id) ->
    {
      final Fragment f = getFragmentForPosition(position, mWear);
      mA.getFragmentManager().beginTransaction()
                            .replace(R.id.dataFeedContents, f)
                            .commit();
      list.setItemChecked(position, true);
      drawer.closeDrawer(list);
      mCurrentlyDisplayedFragmentIndex = position;
    };
    list.setOnItemClickListener(listener);

    final Bundle icicle = args.icicle;
    mCurrentlyDisplayedFragmentIndex = null == icicle ? 0 : icicle.getInt(LAST_OPEN_FRAGMENT_INDEX);
    listener.onItemClick(null, null, mCurrentlyDisplayedFragmentIndex, 0); // Switch to the initial fragment

    final Toolbar toolbar = mA.findViewById(R.id.dataFeedToolbar);
    toolbar.setTitle(R.string.data_feed_title);
    mDrawerToggle = new ActionBarDrawerToggle(mA, drawer, toolbar, R.string.drawer_open_desc, R.string.drawer_closed_desc);
  }

  public void onSaveInstanceState(@NonNull Bundle instanceState)
  {
    instanceState.putInt(LAST_OPEN_FRAGMENT_INDEX, mCurrentlyDisplayedFragmentIndex);
  }

  public void onRequestPermissionsResult(final int requestCode, final String[] permissions, final int[] results)
  {
    if (0 != permissions.length) startGeofenceService(mA);
  }

  // Handling the drawer.
  public void onConfigurationChanged(final Configuration c)
  {
    mDrawerToggle.onConfigurationChanged(c);
  }

  public boolean onOptionsItemSelected(final MenuItem i)
  {
    return mDrawerToggle.onOptionsItemSelected(i);
  }

  public void onPostCreate(final Bundle b)
  {
    mDrawerToggle.syncState();
    if (PackageManager.PERMISSION_GRANTED == ContextCompat.checkSelfPermission(mA, Manifest.permission.ACCESS_FINE_LOCATION))
      startGeofenceService(mA);
    else
      ActivityCompat.requestPermissions(mA, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1);
  }

  @Nullable private static FragmentWrapper<?> getFragmentForPosition(final int position, final Wear wear)
  {
    switch (position)
    {
      case 0 : return new FragmentWrapper<>(MessagesFragment.class, wear);
      case 1 : return new FragmentWrapper<>(ImageSelectorFragment.class, wear);
      case 2 : return new FragmentWrapper<>(LogsAndDataFragment.class, wear);
      case 3 : return new FragmentWrapper<>(DebugToolsFragment.class, wear);
    }
    return null;
  }

  private void startGeofenceService(final Activity activity)
  {
    final Intent i = new Intent(activity, GeofenceTransitionReceiverService.class);
    i.setAction(GeofenceTransitionReceiver.ACTION_MANUAL_START);
    activity.startService(i);
  }
}

File: Feed/src/main/java/com/j/jface/feed/JFaceDataFeed.kt
package com.j.jface.feed

import android.Manifest
import android.app.Activity
import android.content.Context.MODE_PRIVATE
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.os.Bundle
import android.view.MenuItem
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.lifecycleScope
import com.j.jface.R
import com.j.jface.feed.fragments.DebugToolsFragment
import com.j.jface.feed.fragments.ImageSelectorFragment
import com.j.jface.feed.fragments.LogsAndDataFragment
import com.j.jface.feed.fragments.MessagesFragment
import com.j.jface.lifecycle.ActivityWrapper
import com.j.jface.lifecycle.AuthTrampoline
import com.j.jface.lifecycle.FragmentWrapper
import com.j.jface.lifecycle.WrappedActivity
import com.j.jface.org.editor.AuthTrampolineTodoEditorBoot
import com.j.jface.wear.Wear

@Suppress("UNUSED_ANONYMOUS_PARAMETER")

const val OPEN_FRAGMENT_EXTRA = "fragment"
const val LAST_OPEN_FRAGMENT_INDEX = "last_open_fragment_index"

class AuthTrampolineJFaceDataFeedBoot : ActivityWrapper<AuthTrampolineJFaceDataFeed>()
class AuthTrampolineJFaceDataFeed(args : WrappedActivity.Args) : AuthTrampoline(args) { override val trampolineDestination get() = JFaceDataFeedBoot::class.java }
class JFaceDataFeedBoot : ActivityWrapper<JFaceDataFeed>()
class JFaceDataFeed(args : WrappedActivity.Args) : WrappedActivity(args)
{
  private val mDrawerToggle : ActionBarDrawerToggle
  private val mWear : Wear

  init
  {
    mA.setContentView(R.layout.data_feed_drawer)
    mWear = Wear(mA)
    val drawer = mA.findViewById<DrawerLayout>(R.id.dataFeedDrawer)

    // Android thinks the current tab should not be saved across sessions, but for
    // this app it's incorrect, it should be. The icicle is not passed back when
    // opening from a tap on the icon on the home screen, so the state can't be
    // usefully stored there. Instead, persist it.
    val statePrefs = mA.getSharedPreferences("statePrefs", MODE_PRIVATE)
    val fragments = arrayOf("Messages", "Settings", "Logs & data", "Debug tools")
    val list = mA.findViewById<ListView>(R.id.dataFeedDrawerContents)
    list.adapter = ArrayAdapter(mA, R.layout.data_feed_drawer_item, fragments)
    @Suppress("UNUSED_ANONYMOUS_PARAMETER")
    val listener = AdapterView.OnItemClickListener { parent, view, position, id ->
      val f = getFragmentForPosition(position, mWear)!!
      mA.supportFragmentManager.beginTransaction()
       .replace(R.id.dataFeedContents, f)
       .commit()
      list.setItemChecked(position, true)
      drawer.closeDrawer(list)
      statePrefs.edit().putInt(LAST_OPEN_FRAGMENT_INDEX, position).apply()
    }
    list.onItemClickListener = listener

    val intentRequestedFragment = args.intent?.getStringExtra(OPEN_FRAGMENT_EXTRA)
    val fragmentToDisplay : Int = if (null != intentRequestedFragment) fragments.indexOf(intentRequestedFragment) else statePrefs.getInt(LAST_OPEN_FRAGMENT_INDEX, -1)
    mA.lifecycleScope.launchWhenResumed {
      listener.onItemClick(null, null, if (fragmentToDisplay in 0..3) fragmentToDisplay else 3, 0) // Switch to the initial fragment
    }

    val toolbar = mA.findViewById<Toolbar>(R.id.dataFeedToolbar)
    toolbar.setTitle(R.string.data_feed_title)
    mDrawerToggle = ActionBarDrawerToggle(mA, drawer, toolbar, R.string.drawer_open_desc, R.string.drawer_closed_desc)
  }

  companion object
  {
    fun activityClass() = AuthTrampolineJFaceDataFeedBoot::class.java
  }

  override fun onRequestPermissionsResult(requestCode : Int, permissions : Array<String>, results : IntArray)
  {
    if (permissions.isNotEmpty()) startGeofenceService(mA)
  }

  // Handling the drawer.
  override fun onConfigurationChanged(c : Configuration)
  {
    mDrawerToggle.onConfigurationChanged(c)
  }

  override fun onOptionsItemSelected(i : MenuItem) : Boolean
  {
    return mDrawerToggle.onOptionsItemSelected(i)
  }

  override fun onPostCreate(b : Bundle?)
  {
    mDrawerToggle.syncState()
    if (PackageManager.PERMISSION_GRANTED == ContextCompat.checkSelfPermission(mA, Manifest.permission.ACCESS_FINE_LOCATION))
      startGeofenceService(mA)
    else
      ActivityCompat.requestPermissions(mA, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 1)
  }

  private fun getFragmentForPosition(position : Int, wear : Wear) : FragmentWrapper<*>?
  {
    when (position)
    {
      0 -> return FragmentWrapper(MessagesFragment::class.java, wear)
      1 -> return FragmentWrapper(ImageSelectorFragment::class.java, wear)
      2 -> return FragmentWrapper(LogsAndDataFragment::class.java, wear)
      3 -> return FragmentWrapper(DebugToolsFragment::class.java, wear)
    }
    return null
  }

  private fun startGeofenceService(activity : Activity)
  {
    val i = Intent(activity, GeofenceTransitionReceiverService::class.java)
    i.action = GeofenceTransitionReceiver.ACTION_MANUAL_START
    activity.startService(i)
  }
}
